src/rust/src/lib.rs

```diff
@@ -66,8 +73,130 @@ mod rust_bridge {
 mod b64;
 use b64::{from_base64, to_base64};
 
-mod contract;
-use contract::invoke_host_function;
-
 mod log;
 use crate::log::init_logging;
+
+// The tree of dependencies for the contract host is multi-versioned to reduce
+// the possibility of observable semantic drift over code versions.
+//
+// Formally: assume ledger N is originally executed on-chain by copes of core
+// all running against code version X. Then later assume that we wish to upgrade
+// dependencies to code version Y. We say that drift has occurred if N replays
+// with a different tx result or meta hash under X and Y. We want to ensure that
+// if there is drift between X and Y when replaying N, we catch and compensate
+// for it, either by preserving code version X in core forever and replaying N
+// under it, or at least keeping around a special-case path in code version Y
```

Comment:
> preserving code version X in core forever and replaying N under it
If version X has a security vulnerability we don't want to keep the ability to replay a ledger under this version. 
I think this should need to be handled on a case by case basis depending on the vulnerability.

Reply:
Yeah, I'm cleaning up this language -- we'll need to go case-by-case on any drift.